File: src/Calculator/Controls/CalculatorButton.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include "pch.h"
#include "CalculatorButton.h"
#include "CalcViewModel/Common/CalculatorButtonPressedEventArgs.h"

using namespace CalculatorApp;
using namespace CalculatorApp::Common;
using namespace CalculatorApp::Controls;
using namespace Windows::System;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Input;
using namespace Windows::UI::Xaml::Data;
using namespace Platform;
using namespace Windows::Foundation::Collections;

DEPENDENCY_PROPERTY_INITIALIZATION(CalculatorButton, ButtonId);
DEPENDENCY_PROPERTY_INITIALIZATION(CalculatorButton, AuditoryFeedback);
DEPENDENCY_PROPERTY_INITIALIZATION(CalculatorButton, HoverBackground);
DEPENDENCY_PROPERTY_INITIALIZATION(CalculatorButton, HoverForeground);
DEPENDENCY_PROPERTY_INITIALIZATION(CalculatorButton, PressBackground);
DEPENDENCY_PROPERTY_INITIALIZATION(CalculatorButton, PressForeground);

CalculatorButton::CalculatorButton()
{
    // Set the default bindings for this button, these can be overwritten by Xaml if needed
    // These are a replacement for binding in styles
    Binding^ commandBinding = ref new Binding();
    commandBinding->Path = ref new PropertyPath("ButtonPressed");
    this->SetBinding(Button::CommandProperty, commandBinding);
}

void CalculatorButton::OnKeyDown(KeyRoutedEventArgs^ e)
{
    // Ignore the Enter key
    if (e->Key == VirtualKey::Enter)
    {
        return;
    }

    Button::OnKeyDown(e);
}

void CalculatorButton::OnKeyUp(KeyRoutedEventArgs^ e)
{
    // Ignore the Enter key
    if (e->Key == VirtualKey::Enter)
    {
        return;
    }

    Button::OnKeyUp(e);
}

void CalculatorButton::OnButtonIdPropertyChanged(
    NumbersAndOperatorsEnum /*oldValue*/, 
    NumbersAndOperatorsEnum newValue)
{
    this->CommandParameter = ref new CalculatorButtonPressedEventArgs(AuditoryFeedback, newValue);
}

void CalculatorButton::OnAuditoryFeedbackPropertyChanged(
    String^ /*oldValue*/,
    String^ newValue)
{
    this->CommandParameter = ref new CalculatorButtonPressedEventArgs(newValue, ButtonId);
}

File: src/Calculator.Shared/Controls/CalculatorButton.cs
using CalculatorApp.Common;
using System;
using System.Collections.Generic;
using System.Text;
using Windows.System;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Data;
using Windows.UI.Xaml.Input;
using Windows.UI.Xaml.Media;

namespace CalculatorApp.Controls
{
	public partial class CalculatorButton : Button
	{
		public CalculatorButton()
		{
			// Set the default bindings for this button, these can be overwritten by Xaml if needed
			// These are a replacement for binding in styles
			Binding commandBinding = new Binding();
			commandBinding.Path = new PropertyPath("ButtonPressed");
			this.SetBinding(Button.CommandProperty, commandBinding);
		}

		// ButtonId
		public NumbersAndOperatorsEnum ButtonId
		{
			get { return (NumbersAndOperatorsEnum)GetValue(ButtonIdProperty); }
			set { SetValue(ButtonIdProperty, value); }
		}

		public static readonly DependencyProperty ButtonIdProperty =
			DependencyProperty.Register(
                "ButtonId",
                typeof(NumbersAndOperatorsEnum), 
                typeof(CalculatorButton),
                new PropertyMetadata(
                    null, 
                    (s, e) => (s as CalculatorButton)?.OnButtonIdPropertyChanged(
                        (NumbersAndOperatorsEnum)(e.OldValue ?? NumbersAndOperatorsEnum.None),
                        (NumbersAndOperatorsEnum)(e.NewValue ?? NumbersAndOperatorsEnum.None)
                    )
                )
            );

		// AuditoryFeedback
		public string AuditoryFeedback
		{
			get { return (string)GetValue(AuditoryFeedbackProperty); }
			set { SetValue(AuditoryFeedbackProperty, value); }
		}

		public static readonly DependencyProperty AuditoryFeedbackProperty =
			DependencyProperty.Register("AuditoryFeedback", typeof(string), typeof(CalculatorButton), new PropertyMetadata(null));

		// HoverBackground
		public Brush HoverBackground
		{
			get { return (Brush)GetValue(HoverBackgroundProperty); }
			set { SetValue(HoverBackgroundProperty, value); }
		}

		public static readonly DependencyProperty HoverBackgroundProperty =
			DependencyProperty.Register("HoverBackground", typeof(Brush), typeof(CalculatorButton), new PropertyMetadata(null));

		// HoverForeground
		public Brush HoverForeground
		{
			get { return (Brush)GetValue(HoverForegroundProperty); }
			set { SetValue(HoverForegroundProperty, value); }
		}

		public static readonly DependencyProperty HoverForegroundProperty =
			DependencyProperty.Register("HoverForeground", typeof(Brush), typeof(CalculatorButton), new PropertyMetadata(null));

		// PressBackground
		public Brush PressBackground
		{
			get { return (Brush)GetValue(PressBackgroundProperty); }
			set { SetValue(PressBackgroundProperty, value); }
		}

		public static readonly DependencyProperty PressBackgroundProperty =
			DependencyProperty.Register("PressBackground", typeof(Brush), typeof(CalculatorButton), new PropertyMetadata(null));

		// PressForeground
		public Brush PressForeground
		{
			get { return (Brush)GetValue(PressForegroundProperty); }
			set { SetValue(PressBackgroundProperty, value); }
		}

		public static readonly DependencyProperty PressForegroundProperty =
			DependencyProperty.Register("PressForeground", typeof(Brush), typeof(CalculatorButton), new PropertyMetadata(null));

		// PROTECTED 

		protected override void OnKeyDown(KeyRoutedEventArgs e)
		{
			// Ignore the Enter key
			if (e.Key == VirtualKey.Enter)
			{
				return;
			}

			base.OnKeyDown(e);
		}

        protected override void OnKeyUp(KeyRoutedEventArgs e)
		{
			// Ignore the Enter key
			if (e.Key == VirtualKey.Enter)
			{
				return;
			}

            base.OnKeyUp(e);
		}

		// PRIVATE

		private void OnButtonIdPropertyChanged(NumbersAndOperatorsEnum oldValue, NumbersAndOperatorsEnum newValue)
		{
			this.CommandParameter = new CalculatorButtonPressedEventArgs(AuditoryFeedback, newValue);
		}

		private void OnAuditoryFeedbackPropertyChanged(string oldValue, string newValue)
		{
			this.CommandParameter = new CalculatorButtonPressedEventArgs(newValue, ButtonId);
		}
	}
}
